runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/FlinkBatchPipelineTranslator.java

```diff
@@ -57,54 +52,38 @@ public FlinkBatchPipelineTranslator(ExecutionEnvironment env, PipelineOptions op
   // --------------------------------------------------------------------------------------------
 
   @Override
-  public void enterCompositeTransform(TransformTreeNode node) {
+  public CompositeBehavior enterCompositeTransform(TransformTreeNode node) {
     LOG.info(genSpaces(this.depth) + "enterCompositeTransform- " + formatNodeName(node));
 
     PTransform<?, ?> transform = node.getTransform();
-    if (transform != null && currentCompositeTransform == null) {
 
+    if (transform != null) {
       BatchTransformTranslator<?> translator = FlinkBatchTransformTranslators.getTranslator(transform);
       if (translator != null) {
-        currentCompositeTransform = transform;
         if (transform instanceof CoGroupByKey && node.getInput().expand().size() != 2) {
-          // we can only optimize CoGroupByKey for input size 2
-          currentCompositeTransform = null;
+          return CompositeBehavior.ENTER_TRANSFORM;
         }
+        applyBatchTransform(transform, node, translator);
+        LOG.info(genSpaces(this.depth) + "leaveCompositeTranform-" + formatNodeName(node));
```

Comment:
True. I wasn't sure what kind of message you might like to see here so I just made the begin/end symmetrical. Maybe something like "(not recursing) enterCompositeTransform" or whatever would work. But then the format of the string made me think maybe you wanted it to be machine parseable.

I am happy to do anything with these log messages and the `depth`.